### ISSUE TITLE
PermissionManager Transactional Propagation.NEVER

### ISSUE THREAD

**Issue by anthonysena** [OHDSI/WebAPI#1686]
Our WebAPI instance experienced an issue (root cause unknown) whereby the following unhandled exception was thrown over-and-over again:

````
ERROR https-jsse-nio-8443-exec-715 org.ohdsi.webapi.shiro.filters.ExceptionHandlerFilter -  - Error during filtering
org.springframework.transaction.IllegalTransactionStateException: Existing transaction found for transaction marked with propagation 'never'
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.handleExistingTransaction(AbstractPlatformTransactionManager.java:409)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.getTransaction(AbstractPlatformTransactionManager.java:353)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.createTransactionIfNecessary(TransactionAspectSupport.java:461)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:277)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:671)
	at org.ohdsi.webapi.shiro.PermissionManager$$EnhancerBySpringCGLIB$$5c7cc510.clearAuthorizationInfoCache(<generated>)
	at org.ohdsi.webapi.shiro.filters.CacheFilter.doFilter(CacheFilter.java:33)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.ohdsi.webapi.shiro.filters.ExceptionHandlerFilter.doFilter(ExceptionHandlerFilter.java:28)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:109)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:197)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.boot.web.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:130)
	at org.springframework.boot.web.support.ErrorPageFilter.access$000(ErrorPageFilter.java:66)
	at org.springframework.boot.web.support.ErrorPageFilter$1.doFilterInternal(ErrorPageFilter.java:105)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.boot.web.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:123)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:198)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:478)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:624)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:80)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:624)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:799)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:861)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1455)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Unknown Source)
````
It appears that when this exception is thrown, there is no recovery and the process gets stuck in an infinite loop until the thread is destroyed (in our case by restarting Tomcat). 

In reviewing the stack trace, this `@Transactional(propagation = Propagation.NEVER)` annotation appears to be related: https://github.com/OHDSI/WebAPI/blob/master/src/main/java/org/ohdsi/webapi/shiro/PermissionManager.java#L139. Keep in mind, the instance of WebAPI that experienced this error does not have security enabled so it is unclear (to me at least) why this is invoked. 

This annotation was introduced when fixing the `SecurityIT` tests: https://github.com/OHDSI/WebAPI/pull/1311

**Comment by anthonysena** [comment]
Just adding to this issue as we again faced this issue today. The exception above appears to happen after checking to see if a cohort has an existing cache entry. 

````
2020-12-08 01:43:49.868 INFO taskExecutor-6 org.springframework.batch.core.launch.support.SimpleJobLauncher -  - Job: [SimpleJob: [name=irAnalysis]] launched with the following parameters: [{jobName=IR Analysis: 505: CDM (v1294) (CDM_CDM), analysis_id=505, source_id=274, sessionId=gb27r392, target_table=temp_cohort_gb27r392, time=1607391829824, jobAuthor=anonymous}]
2020-12-08 01:43:50.027 INFO taskExecutor-6 org.springframework.batch.core.job.SimpleStepHandler -  - Executing step: [irAnalysis.createCohortTable]
2020-12-08 01:43:50.814 INFO taskExecutor-6 org.springframework.batch.core.job.SimpleStepHandler -  - Executing step: [irAnalysis.generateCohort]
2020-12-08 01:43:53.149 INFO pool-327-thread-1 org.ohdsi.webapi.generationcache.GenerationCacheHelper -  - Computes cache if absent for type = COHORT, design = 1719005750, source id = 274
2020-12-08 01:43:53.161 INFO pool-327-thread-1 org.ohdsi.webapi.generationcache.GenerationCacheHelper -  - Retrieves or invalidates cache for cohort id = 11071
2020-12-08 01:43:54.156 INFO pool-327-thread-1 org.ohdsi.webapi.generationcache.CohortGenerationCacheProvider -  - Checksum of Generation cache for designHash = 1719005750 has been calculated in 989 milliseconds
2020-12-08 01:43:54.156 INFO pool-327-thread-1 org.ohdsi.webapi.generationcache.GenerationCacheHelper -  - Using cached generation results for COHORT with id=11071 and source=CDM_CDM
2020-12-08 01:43:54.156 INFO pool-327-thread-1 org.ohdsi.webapi.generationcache.GenerationCacheHelper -  - Finished computation cache if absent for cohort id = 11071
2020-12-08 01:43:58.857 INFO pool-327-thread-1 org.ohdsi.webapi.generationcache.GenerationCacheHelper -  - Computes cache if absent for type = COHORT, design = -1398439376, source id = 274
2020-12-08 01:43:58.871 INFO pool-327-thread-1 org.ohdsi.webapi.generationcache.GenerationCacheHelper -  - Retrieves or invalidates cache for cohort id = 11073
2020-12-08 01:43:59.875 INFO pool-327-thread-1 org.ohdsi.webapi.generationcache.CohortGenerationCacheProvider -  - Checksum of Generation cache for designHash = -1398439376 has been calculated in 998 milliseconds
2020-12-08 01:43:59.875 INFO pool-327-thread-1 org.ohdsi.webapi.generationcache.GenerationCacheHelper -  - Using cached generation results for COHORT with id=11073 and source=CDM_CDM
2020-12-08 01:43:59.875 INFO pool-327-thread-1 org.ohdsi.webapi.generationcache.GenerationCacheHelper -  - Finished computation cache if absent for cohort id = 11073
2020-12-08 01:44:00.750 ERROR https-jsse-nio-8443-exec-28 org.ohdsi.webapi.shiro.filters.ExceptionHandlerFilter -  - Error during filtering
org.springframework.transaction.IllegalTransactionStateException: Existing transaction found for transaction marked with propagation 'never'
...
<see full exception message above>
...
````

Looking at [GenerationCacheHelper.java#L49](https://github.com/OHDSI/WebAPI/blob/master/src/main/java/org/ohdsi/webapi/generationcache/GenerationCacheHelper.java#L49), this code has a comment mentioning that this code block is executed in a separate transaction. Could this be where the conflict exists whereby the ` @Transactional(propagation = Propagation.NEVER)` annotation is then in conflict since this operation is happening inside of a transaction that requires propagation?

**Comment by anthonysena** [comment]
Closed via #1715
